src/aleph/storage.py

```diff
@@ -22,113 +24,207 @@
         if "item_content" not in message:
-            LOGGER.warning(f"No item_content in message {message.get('item_hash')}")
-            return -1, 0  # never retry, bogus data
+            error_msg = f"No item_content in message {message.get('item_hash')}"
+            LOGGER.warning(error_msg)
+            raise InvalidContent(error_msg)  # never retry, bogus data
         try:
             item_content = await json_async_loads(message["item_content"])
-        except (json.JSONDecodeError, KeyError):
-            LOGGER.exception("Can't decode JSON")
-            return -1, 0  # never retry, bogus data
-        return item_content, len(message["item_content"])
+        except (json.JSONDecodeError, KeyError) as e:
+            error_msg = f"Can't decode JSON: {e}"
+            LOGGER.exception(error_msg)
+            raise InvalidContent(error_msg)
+
+        return JsonContent(
+            hash=item_hash,
+            source=ContentSource.INLINE,
+            value=item_content,
+            raw_content=message["item_content"],
+        )
     else:
-        LOGGER.exception("Unknown item type: %s", item_type)
-        return (
-            None,
-            0,
-        )  # unknown, could retry later? shouldn't have arrived this far though.
+        # unknown, could retry later? shouldn't have arrived this far though.
+        error_msg = f"Unknown item type: '{item_type}'."
+        raise ContentCurrentlyUnavailable(error_msg)
+
+
+async def fetch_content_from_network(
+    content_hash: str, engine: ItemType, timeout: int
+) -> Optional[bytes]:
+    content = None
+    enabled_clients = app["config"].p2p.clients.value
+
+    if "protocol" in enabled_clients:
+        streamer = get_streamer()
+        content = await streamer.request_hash(content_hash)
+
+    if content is None and "http" in enabled_clients:
+        content = await p2p_http_request_hash(content_hash, timeout=timeout)
+
+    if content is not None:
+        await verify_content_hash(content, engine, content_hash)
+
+    return content
+
+
+async def compute_content_hash_ipfs(
+    content: bytes, expected_hash: str
+) -> Optional[str]:
+    """
+    Computes the IPFS hash of the content.
+    :param content: Content to hash.
+    :param expected_hash: Expected hash of the content.
+    :return: The computed hash of the content. Can return None if the operation fails for some reason.
+    """
+
+    # TODO: get a better way to compare hashes (without depending on the IPFS daemon)
+    try:
+        cid_version = 0
+        if len(expected_hash) >= 58:
+            cid_version = 1
+
+        computed_hash = await add_ipfs_bytes(content, cid_version=cid_version)
+        return computed_hash
+
+    except asyncio.TimeoutError:
+        LOGGER.warning(f"Timeout while computing IPFS hash for '{expected_hash}'.")
+        return None
+
+
+async def compute_content_hash_sha256(content: bytes, expected_hash: str) -> str:
+    return get_sha256(content)
```

Comment:
Should this call run in a thread (`asyncio.run_in_executor`) ?

Reply:
To be honest I don't really see a point of running computing-intensive tasks in the executor, at least if we use the `ThreadPoolExecutor`. It just adds overhead and no parallelization whatsoever.

If you want we can investigate creating a task pool for CPU-bound tasks running on a `ProcessPoolExecutor`, but I'm really doubtful about the efficiency for most messages.